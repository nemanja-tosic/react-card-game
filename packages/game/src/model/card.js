// @flow
export class Card {
  code: string;
  suit: string;
  value: number;

  constructor(code: string, suit: string, value: number) {
    this.code = code;
    this.suit = suit;
    this.value = value;
  }

  takes(card: Card) {
    return this.value >= card.value;
  }
}
