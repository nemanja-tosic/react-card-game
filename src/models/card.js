// @flow

export class Card {
  code: string;
  imageUrl: string;

  constructor(code: string) {
    this.code = code;
    this.imageUrl = `https://deckofcardsapi.com/static/img/${code}.png`;
  }
}
