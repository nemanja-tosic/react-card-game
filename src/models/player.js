// @flow
import { Card } from './card';

export class Player {
  id: number;
  name: string;
  cards: Array<Card> = [];
  score: number = 0;
  isActive = false;
  takenCards: Array<Array<Card>> = [];
  isUserControlled: boolean;

  constructor(id: number, name: string, cards: Array<Card>, takenCards: Array<Array<Card>>, score: number, isActive: boolean, isUserControlled: boolean) {
    this.id = id;
    this.name = name;
    this.cards = cards;
    this.takenCards = takenCards;
    this.score = score;
    this.isActive = isActive;
    this.isUserControlled = isUserControlled;
  }
}
