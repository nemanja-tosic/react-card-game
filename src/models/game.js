// @flow
import { Card } from './card';
import { Player } from './player';

type GameState = 'notStarted' | 'loading' | 'inProgress' | 'over';

export class Game {
  players: Array<Player>;
  talon: Array<Card>;
  state: GameState;
  winners: Array<Player>;

  constructor(players: Array<Player>, talon: Array<Card>, state: GameState, winners: Array<Player>) {
    this.players = players;
    this.talon = talon;
    this.state = state;
    this.winners = winners;
  }
}
