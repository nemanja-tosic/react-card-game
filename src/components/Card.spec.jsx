// @flow
import React from 'react';
import { shallow } from 'enzyme';

import { Card } from '../models/card';
import { TestCards } from '../test/support/test-cards';
import CardComponent from './Card';

describe('Card', () => {
  let component;
  let spy;
  let cardRefSpy;

  const card: Card = TestCards.create();

  beforeEach(() => {
    spy = jest.fn();
    cardRefSpy = jest.fn();
    component = shallow(<CardComponent card={card} onClick={spy} cardRef={cardRefSpy}/>);
  });

  describe('when the faceDown prop is true', () => {
    beforeEach(() => {
      component.setProps({faceDown: true});
    });

    describe('and the orientation is landscape', () => {
      it('should display the face down card image in the landscape orientation', () => {
        component.setProps({orientation: 'landscape'});

        expect(component.getNode()).toMatchSnapshot();
      });
    });

    describe('and the orientation is portrait', () => {
      it('should display the face down card image in the portrait orientation', () => {
        component.setProps({orientation: 'portrait'});

        expect(component.getNode()).toMatchSnapshot();
      });
    });

    describe('and the orientation is not provided', () => {
      it('should display the face down card image in the portrait orientation', () => {
        expect(component.getNode()).toMatchSnapshot();
      });
    });
  });

  describe('when the faceDown prop is false', () => {
    beforeEach(() => {
      component.setProps({faceDown: false});
    });

    it('should display the card image', () => {
      expect(component.getNode()).toMatchSnapshot();
    });
  });

  describe('when the card is clicked', () => {
    it('should call the onClick prop if present', () => {
      component.first().simulate('click');

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should not call the onClick prop if not present', () => {
      component.setProps({onClick: undefined});

      component.first().simulate('click');

      expect(spy).not.toHaveBeenCalled();
    });
  });
});
