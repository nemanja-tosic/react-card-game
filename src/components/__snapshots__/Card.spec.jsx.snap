// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card when the faceDown prop is false should display the card image 1`] = `
<div
  className="Card"
  onClick={[Function]}
>
  <div
    className="Card__container Card__container--portrait"
  >
    <div
      className="Card__image"
      style={
        Object {
          "backgroundImage": "url(https://deckofcardsapi.com/static/img/JS.png)",
        }
      }
    />
  </div>
</div>
`;

exports[`Card when the faceDown prop is true and the orientation is landscape should display the face down card image in the landscape orientation 1`] = `
<div
  className="Card"
  onClick={[Function]}
>
  <div
    className="Card__container Card__container--landscape"
  >
    <div
      className="Card__image"
      style={
        Object {
          "backgroundImage": "url(facedown-landscape.png)",
        }
      }
    />
  </div>
</div>
`;

exports[`Card when the faceDown prop is true and the orientation is not provided should display the face down card image in the portrait orientation 1`] = `
<div
  className="Card"
  onClick={[Function]}
>
  <div
    className="Card__container Card__container--portrait"
  >
    <div
      className="Card__image"
      style={
        Object {
          "backgroundImage": "url(facedown-portrait.png)",
        }
      }
    />
  </div>
</div>
`;

exports[`Card when the faceDown prop is true and the orientation is portrait should display the face down card image in the portrait orientation 1`] = `
<div
  className="Card"
  onClick={[Function]}
>
  <div
    className="Card__container Card__container--portrait"
  >
    <div
      className="Card__image"
      style={
        Object {
          "backgroundImage": "url(facedown-portrait.png)",
        }
      }
    />
  </div>
</div>
`;
